### PR TITLE
 Add and use a Builder for the DataStream class

### PR DESCRIPTION
**Pull request by nielsbauman**
This serves as preparation for a (potential) upcoming refactoring of the `DataStream` class fields, but it definitely still has value if that refactoring doesn't happen as it removes some boilerplate/repetitive code in several places.

### CODE REVIEW

File: modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java
Comment (nielsbauman) [inline modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java:143]: In most places, I added `setRolloverOnWrite(false)`, because that's what the original constructor did there. I'm guessing it doesn't actually need to be set in a lot of those places, though. What are people's opinion on (me) taking some time to go through these changes and check on a per-case basis whether it actually needs to be set explicitly?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:1317]: If I make a new instance builder and set no values on anything and try to call build(), would we get a valid data stream object out of it? What would the generation be in that case?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:1332]: I would check out how something like IndexMetadata works with its builder - it has required parameters in the constructor if not copying an existing instance.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java:202]: Hmm, not related to this specific PR, and maybe someone else can clarify, but it seems like we're very regularly clearing the rollover on write flag. Is that intentional? I can see it making sense in a lot of the other places, but I wouldn't have expected it being cleared when updating a lifecycle.

**Review by jbaiera — CHANGES_REQUESTED** [review]
Looking very promising - it's funny how much gets simplified with this kind of change.

I left a couple comments for some possible improvements and some open questions regarding the rollover on write switches.

File: modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java
Comment (jbaiera) [reply modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java:143]: I have a general concern that because the setting was being set under all these different constructors that maybe there are some cases where we're unintentionally clearing the flag. I pointed one out that I was perplexed by in my review. Maybe @gmarouli can weigh in on this?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:1332]: I think I was a bit quick with adding the empty constructor here (force of habit). On second thought (or maybe first thought 🙃), it might actually make sense to only have the "copy" constructor for now, as we don't have any other use cases currently for using a builder to construct a `DataStream` from scratch. We don't have a lot of optional fields and, looking at the usages of the `DataStream` constructors, it doesn't seem to add any value to have defaults for fields. Would you be okay with removing this constructor altogether? That would also resolve your other question on fields not being set.

File: modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java
Comment (gmarouli) [reply modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java:143]: You are absolutely right, this is definitely a bug, and I need to look into this, it should not clear the flag unless it was rolled over.

Comment (gmarouli) [reply modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java:143]: We can split the PRs, I can work on it after this is merged.

Comment (gmarouli) [reply modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java:143]: Actually, I will fix it first so I can backport it easier to 8.13

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:1332]: I think the main things we would need to always include would be the name and the first write index. Everything else could have a sensible default based on the current creation logic:

```
name: Required
indices: Required

generation: Default 1L
metadata: Default null
hidden: Default false
replicated: Default false
system: Default false
allowCustomRouting: Default false
indexMode: Default null
lifecycle: Default null
failureStore: Default false
failureIndices: Default empty list
rolloverOnWrite: Default false
autoShardingEvent: Default null
```

File: modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java
Comment (nielsbauman) [reply modules/data-streams/src/test/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeServiceTests.java:143]: See https://github.com/elastic/elasticsearch/pull/107122 for a fix.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java:202]: Discussed and addressed in the other comment, so resolving this one.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:1317]: Discussed and addressed in the comment below -- resolving this one.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java
Comment (gmarouli) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java:1781]: Nit: I like having a static method for creating a new builder, calling the constructor then becomes `DataStream.builder(original)`.

But this is just a preference. In the code we use both patterns as far as I know.

Comment (nielsbauman) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java:1781]: Any reasoning for why you prefer that? Out of curiosity :)

Comment (gmarouli) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java:1781]: Of course, 
- I find it nice in IDE to be able to write `DataStream.` and immediately find the builder constructor. This way, I check if there is a builder and use it in one step. Otherwise, I see there is a `DataStream.Builder` and then I have to go back and write `new`
- I think it's more concise to write `DataStream.builder(original)` than `new DataStream.Builder(original)`.

But again, I understand this is my process and my taste, so I am just offering it as an alternative.

Comment (nielsbauman) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java:1781]: Ah, right, okay. Can I propose another option 😉:
```java
public DataStream.Builder builder() {
    return new DataStream.Builder(this);
}
```
I think I'd prefer that over `DataStream.builder(original)`, as the latter is basically just an alias of the builder's constructor, so then I see slightly less value. But, being able to create a builder straight from a data stream instance for copying purposes, seems like a valuable "shortcut". What do you think?

Comment (gmarouli) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java:1781]: I like the idea in general, not sure if we do this in other places, definitely worth to follow it up.

File: server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java
Comment (gmarouli) [inline server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java:211]: So, here you change the comment from `unsure` to `ensure`. Honestly, I think this comment should be removed because I do not see what is its value. However, I think `ensure` is wrong because `indexMode` is set by:
```
IndexMode indexMode = randomBoolean() ? IndexMode.STANDARD : null
```

Also I think this could be rewritten simpler to:

```
new DataStream.Builder(DataStreamTestHelper.randomInstance()).setReplicated(false).setIndexMode(indexMode).build()
```

File: server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java
Comment (gmarouli) [inline server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java:226]: Similarly here:
```
DataStream ds = new DataStream.Builder(DataStreamTestHelper.randomInstance()).setReplicated(false).setIndexMode(IndexMode.TIME_SERIES).build();
```

**Review by gmarouli — COMMENTED** [review]
@nielsbauman this looks great, I think it fits very well with the way we use `DataStream`. I left some minor comments and opened a discussion about a bug that we can fix at another PR but it was highlighted here because of your builder :) .

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:511]: Ah yeah you're right, good catch! I'll open up a separate PR for this. I'm not sure if I'll be able to get lazy rollover completely in before the 8.14 feature freeze, so I'll do a dedicated PR for this.

Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:513]: See https://github.com/elastic/elasticsearch/pull/107247

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java
Comment (nielsbauman) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java:1781]: While searching for `new Builder(this)`, I only get two results: https://github.com/elastic/elasticsearch/blob/9f3cbc120033e869a18349f79e9b6c107fbc5e22/x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/common/http/HttpRequest.java#L294-L297
https://github.com/elastic/elasticsearch/blob/9f3cbc120033e869a18349f79e9b6c107fbc5e22/server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java#L358-L360
I also see things like:
https://github.com/elastic/elasticsearch/blob/9f3cbc120033e869a18349f79e9b6c107fbc5e22/server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java#L440-L442
but that feels like a slightly different use case/scenario. Do you have any preference for the method name? Both `toBuilder` and `copy` work for me. Or is your opinion that we shouldn't do it because it's not widely done in the project?

Comment (nielsbauman) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java:1781]: Yeah I've seen that pattern in several places. Does your motivation for wanting such a static method align with what Mary mentioned? Or are there other reasons you'd like to have such a method (other than that it's seen in other places)? As I said before, I see relatively little value in these methods, as they, IMO, are only marginally more concise than the constructor itself. OTOH, they're also not that verbose so it's definitely not something I'm going to block this PR over for much longer.

**Review by jbaiera — APPROVED** [review]
LGTM (pending green CI). I threw a couple fast suggestions on for validating some important state, but otherwise I think it should be good to go at this point. Thanks for iterating on this @nielsbauman

**Comment by nielsbauman** [comment]
I removed one more constructor from `DataStream` and used the builder instead (see [f420f93](https://github.com/elastic/elasticsearch/pull/107035/commits/f420f93279e1b90cec38d812ae0dd58664cc5cac)).

**Review by gmarouli — APPROVED** [review]
LGTM, thank you for putting all this effort @nielsbauman! 🥳 435 lines less and the code is more readable, how cool is that!

**Comment by nielsbauman** [comment]
@elasticmachine update branch

**Comment by nielsbauman** [comment]
@elasticmachine update branch

**Comment by nielsbauman** [comment]
@elasticmachine update branch

**Pull request merged**
